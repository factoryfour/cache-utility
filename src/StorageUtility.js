class StorageUtility {
	constructor(config) {
		// Initialize lastChange as when first constructed
		this.lastChange = Date.now();
		// Determine the target storage entity on the window
		this.target = window[config.target];
		// Store the desired tiers
		this.sortedTiers = config.tiers.sort((a, b) => {
			if (a.expiration < b.expiration) {
				return 1;
			}
			if (a.expiration > b.expiration) {
				return -1;
			}
			return 0;
		});
		/* Now that the tiers are sorted in increasing duration, we can
		generate the map of tier name to key. */
		this.tierMap = {};
		this.sortedTiers.forEach((tier, index) => {
			if (index === 0) {
				this.tierMap[tier.name] = {
					key: tier.name,
					expiration: tier.expiration,
				};
			} else {
				this.tierMap[tier.name] = {
					key: `${this.tierMap[this.sortedTiers[index - 1].name].key}-${tier.name}`,
					expiration: tier.expiration,
				};
			}
		});
	}

	/** Retrieves the associated data from storage */
	get(key, tier) {
		// If the desired tier doesn't exist, print a warning to the console
		if (!this.tierMap[tier]) {
			console.warn(`Tier ${tier} does not exist`);
			return null;
		}

		const now = Date.now();
		const inactivity = now - this.lastChange;
		// Update lastChange
		this.lastChange = now;
		// Initiate the removal of invalid tiers
		this.invalidate(inactivity);
		// If tier is invalid, return null else retrieve
		return this.tierMap[tier].expiration > inactivity ?
			JSON.parse(this.target.getItem(`${this.tierMap[tier].key}-${key}`)) : null;
	}

	/** Puts a desired value into the target store. Returns null if unsuccessful */
	set(key, value, tier) {
		// If the desired tier doesn't exist, print a warning to the console
		if (!this.tierMap[tier]) {
			console.warn(`Tier ${tier} does not exist`);
			return null;
		}

		const newValue = JSON.stringify(value);
		const newKey = `${this.tierMap[tier].key}-${key}`;

		const now = Date.now();
		const inactivity = now - this.lastChange;
		// Update lastChange
		this.lastChange = now;
		// Initiate the removal of invalid tiers
		this.invalidate(inactivity);

		this.target.setItem(newKey, newValue);
	}

	invalidate(inactivity) {
		/* Based on the amount of inactivity, determine
		which tiers should be removed and remove all of
		those tiers */
		let filter = '';

		const invalid = this.sortedTiers.some((tier) => {
			if (tier.expiration > inactivity) {
				filter = filter.concat(`${tier.name}-`);
				return false;
			}
			filter = filter.concat(`${tier.name}-`);
			return true;
		});
		// If there exists a tier to invalidate, do it
		if (invalid) {
			Object.keys(this.target).forEach((key) => {
				// If filter is present, remove
				if (key.indexOf(filter) > -1) {
					this.target.removeItem(key);
				}
			});
		}
	}
}

export default StorageUtility;
