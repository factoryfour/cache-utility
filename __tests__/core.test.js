import StorageUtility from '../index';

const testTiers = [
	{
		name: 'A',
		expiration: 500,
	},
	{
		name: 'B',
		expiration: 1000,
	},
	{
		name: 'C',
		expiration: 1500,
	},
];

const config = {
	target: 'sessionStorage',
	tiers: testTiers,
};

describe('StorageUtility', () => {
	let ExampleUtility;

	beforeEach(() => {
		ExampleUtility = new StorageUtility(config);
	});

	it('should properly initialize tierMap', () => {
		expect(ExampleUtility.tierMap).toMatchSnapshot();
	});

	it('should properly set a value in target', () => {
		const testValue = { hello: '123' };
		expect(localStorage.length).toBe(0);
		expect(sessionStorage.length).toBe(0);
		ExampleUtility.set('test', testValue, 'A');
		expect(localStorage.length).toBe(0);
		expect(sessionStorage.length).toBe(1);
		expect(sessionStorage.setItem).toHaveBeenCalledWith('C-B-A-test', JSON.stringify(testValue));
		expect(sessionStorage.__STORE__['C-B-A-test']).toMatchSnapshot(); // eslint-disable-line
	});

	it('should properly get a value in target', () => {
		const testValue = { hello: '123' };
		ExampleUtility.set('test', testValue, 'A');
		const result = ExampleUtility.get('test', 'A');
		expect(result).toMatchSnapshot();
	});

	it('should return null when getting from an invalid tier', () => {
		const testValue = { hello: '123' };
		ExampleUtility.set('test', testValue, 'A');
		const result = ExampleUtility.get('test', 'B');
		expect(result).toBe(null);
	});

	it('should return null when attempting to get after the tier has expired and remove key from store', (done) => {
		const testValue = { hello: '123' };
		ExampleUtility.set('test', testValue, 'A');
		setTimeout(() => {
			const result = ExampleUtility.get('test', 'A');
			expect(result).toBe(null);
			expect(sessionStorage.length).toBe(0);
			expect(sessionStorage.removeItem).toHaveBeenCalledWith('C-B-A-test');
			done();
		}, 750);
	});

	it('should persist long term tier over short term tier after get', (done) => {
		ExampleUtility.set('test', { hello: '123' }, 'A');
		ExampleUtility.set('test', { hello: '321' }, 'B');
		expect(sessionStorage.length).toBe(2);
		setTimeout(() => {
			const result = ExampleUtility.get('test', 'B');
			expect(result).toMatchSnapshot();
			expect(sessionStorage.length).toBe(1);
			expect(sessionStorage.removeItem).toHaveBeenCalledWith('C-B-A-test');
			done();
		}, 750);
	});

	it('should persist long term tier over short term tier after set', (done) => {
		ExampleUtility.set('test', { hello: '123' }, 'A');
		ExampleUtility.set('test', { hello: '321' }, 'B');
		expect(sessionStorage.length).toBe(2);
		setTimeout(() => {
			ExampleUtility.set('test', { hello: '456' }, 'C');
			expect(sessionStorage.length).toBe(2);
			expect(sessionStorage.removeItem).toHaveBeenCalledWith('C-B-A-test');
			expect(sessionStorage.__STORE__).toMatchSnapshot(); // eslint-disable-line
			done();
		}, 750);
	});
});
