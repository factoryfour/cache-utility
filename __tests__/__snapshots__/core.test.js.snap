// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageUtility should persist long term tier over short term tier 1`] = `
Object {
  "hello": "321",
}
`;

exports[`StorageUtility should persist long term tier over short term tier after get 1`] = `
Object {
  "hello": "321",
}
`;

exports[`StorageUtility should persist long term tier over short term tier after set 1`] = `
LocalStorage {
  "C-B-test": "{\\"hello\\":\\"321\\"}",
  "C-test": "{\\"hello\\":\\"456\\"}",
}
`;

exports[`StorageUtility should properly get a value in target 1`] = `
Object {
  "hello": "123",
}
`;

exports[`StorageUtility should properly initialize tierMap 1`] = `
Object {
  "A": Object {
    "expiration": 500,
    "key": "C-B-A",
  },
  "B": Object {
    "expiration": 1000,
    "key": "C-B",
  },
  "C": Object {
    "expiration": 1500,
    "key": "C",
  },
}
`;

exports[`StorageUtility should properly set a value in target 1`] = `"{\\"hello\\":\\"123\\"}"`;
